Add `miden-mast-package` crate with `Package` type to represent a compiled Miden program/library.

greenhat opened the pull request:
This PR adds `miden-package` crate with `Package` type ported from the compiler's ad-hoc implementation to represent a compiled Miden program/library. 

Please consider reviewing this PR commit-by-commit. I crafted the commits to be focused on a single change and show the evolution of the implementation.

The reasons for the new crate are explained in https://github.com/0xPolygonMiden/compiler/issues/376 

The compiler's PR that uses `miden-package` from this PR is https://github.com/0xPolygonMiden/compiler/pull/349

plafer commented on package/src/dep/mod.rs:
```diff
@@ -0,0 +1,71 @@
+    /// The standard library
+    Stdlib,
+    /// The base library
+    Miden,
```
Hmmm, the VM doesn't know about `miden-base`'s library, a rollup-specific concept, whereas the VM is technically general-purpose. Is there a way to avoid having this here? Could we consider it to be a user library?

bobbinth replied:
Yeah - I'd rather not put things specific to `miden-base` (and maybe even the standard library) here.

Could we not just use strings and say that `std` or `stdlib` is the standard library and `base` or `miden` is the Miden rollup library?

greenhat replied:
Agree. I probably got carried away in avoiding "stringly typed" approach. Fixed in https://github.com/0xPolygonMiden/miden-vm/pull/1544/commits/ab26b552c6ddf6f4c80e4b94d6b8e466b14b5fc3 
Now, it's just `DependencyName(String)`.

plafer commented on package/src/dep/resolver.rs:
```diff
@@ -0,0 +1,69 @@
+#[derive(Debug, Clone)]
+pub enum DependencyResolution {
+    Local(LocalResolution),
+    // Registry(...),
```
nit: can we remove this? As suggested in a previous comment, we could add a docstring and mention that a future variant could be a registry

greenhat replied:
Sure. Done in https://github.com/0xPolygonMiden/miden-vm/pull/1544/commits/43e0a83488f491dfbe21305dae257336aa287fd8

plafer commented on package/src/dep/resolver.rs:
```diff
@@ -0,0 +1,69 @@
+// ================================================================================================
+
+#[derive(Debug, Clone)]
+pub enum DependencyResolution {
```
Would a better name here be `ResolvedDependency`? I tend to read this as a verb, or "the action to resolve".

Similar comment for `LocalResolution`; maybe could be `LocalResolvedDependency` or something

greenhat replied:
Sure. Done in https://github.com/0xPolygonMiden/miden-vm/pull/1544/commits/43e0a83488f491dfbe21305dae257336aa287fd8

plafer commented on package/src/package.rs:
```diff
@@ -0,0 +1,305 @@
+}
+
+impl Package {
+    const MAGIC: &'static [u8] = b"MASP\0";
+    const FORMAT_VERSION: &'static [u8] = b"1.0\0";
```
nit: could we move those at the top of the file, similar to [how we do with `MastForest`](https://github.com/0xPolygonMiden/miden-vm/blob/ebc8311467833bd34ac3bf0592f857390a37141d/core/src/mast/serialization/mod.rs#L80-L88)? Do we also want the same serialized format for versions as `MastForest` for consistency?

Also maybe for consistency with `MastForest` we'd use version 0 for now?

greenhat replied:
Done.

plafer commented on package/src/package.rs:
```diff
@@ -0,0 +1,305 @@
+    const FORMAT_VERSION: &'static [u8] = b"1.0\0";
+
+    /// Parses a package from the provided bytes
+    pub fn read_from_bytes<B>(bytes: B) -> Result<Self, Report>
+    where
+        B: AsRef<[u8]>,
+    {
+        use alloc::borrow::Cow;
+
+        let bytes = bytes.as_ref();
+
+        let bytes = bytes
+            .strip_prefix(Self::MAGIC)
+            .ok_or_else(|| Report::msg("invalid package: missing header"))?;
+        let bytes = bytes.strip_prefix(Self::FORMAT_VERSION).ok_or_else(|| {
+            Report::msg(format!(
+                "invalid package: incorrect version, expected '1.0', got '{}'",
+                bytes.get(0..4).map(String::from_utf8_lossy).unwrap_or(Cow::Borrowed("")),
+            ))
+        })?;
+
+        bitcode::deserialize(bytes).map_err(Report::msg)
+    }
+
+    /// Serializes the package into a byte array
+    pub fn write_to_bytes(&self) -> Result<Vec<u8>, Report> {
+        let mut bytes = Vec::new();
+        bytes.extend_from_slice(Self::MAGIC);
+        bytes.extend_from_slice(Self::FORMAT_VERSION);
+        let mut data = bitcode::serialize(self).map_err(Report::msg)?;
+        bytes.append(&mut data);
+        Ok(bytes)
+    }
```
Could we implement `Serializable` and `Deserializable` instead of custom functions (similar to `MastForest`)?

greenhat replied:
I wanted to do it initially, but `bitcode` serialization being fallible was the major obstacle.

greenhat replied:
Removed `bitcode`(`serde`) in favor of `Serializable`/`Deserializable`.

plafer commented on package/src/package.rs:
```diff
@@ -0,0 +1,305 @@
+            .ok_or_else(|| Report::msg("invalid package: missing header"))?;
+        let bytes = bytes.strip_prefix(Self::FORMAT_VERSION).ok_or_else(|| {
+            Report::msg(format!(
+                "invalid package: incorrect version, expected '1.0', got '{}'",
```
small nit: we could use `FORMAT_VERSION` instead of hardcoding `1.0`

greenhat replied:
Done.

plafer commented on package/src/package.rs:
```diff
@@ -0,0 +1,305 @@
+        let mut bytes = Vec::new();
+        bytes.extend_from_slice(Self::MAGIC);
+        bytes.extend_from_slice(Self::FORMAT_VERSION);
+        let mut data = bitcode::serialize(self).map_err(Report::msg)?;
```
`bitcode` is a "Rust-only" serialization format that isn't stable across major version. This is probably fine for now, but we'd probably want to move towards a format that anyone can write a deserializer for

greenhat replied:
Good point! Please see my comment https://github.com/0xPolygonMiden/miden-vm/pull/1544#discussion_r1918897284 below.

plafer commented on core/src/program.rs:
```diff
@@ -53,6 +54,16 @@ impl Program {
         Self { mast_forest, entrypoint, kernel }
     }
+
+    /// Produces a new program with the existing [`MastForest`] and provided advice map.
```
```suggestion
    /// Produces a new program with the existing [`MastForest`] and where all key/values in the
    /// provided advice map are added to the internal advice map.
```

plafer commented on assembly/src/library/mod.rs:
```diff
@@ -92,6 +91,16 @@ impl Library {
         Ok(Self { digest, exports, mast_forest })
     }
+
+    /// Produces a new library with the existing [`MastForest`] and provided advice map.
```
```suggestion
    /// Produces a new library with the existing [`MastForest`] and where all key/values in the  
    /// provided advice map are added to the internal advice map.
```

plafer commented on assembly/src/ast/procedure/name.rs:
```diff
@@ -328,3 +333,47 @@ impl FromStr for ProcedureName {
+
+    #[test]
+    fn procedure_name_serialization_rustc_mangled() {
+        // Tests that rustc mangled symbols are serialized and deserialized
+        // see https://doc.rust-lang.org/rustc/symbol-mangling/v0.html#symbol-grammar-summary
```
Why do we need to test this explicitly?

greenhat replied:
I put it as a regression test in case an incompatible validation is (re-)introduced in the `ProcedureName` deserialization.

greenhat replied:
I removed this test since this case is covered in the serialization roundtrip test via using these characters for one of the procedure names in `Arbitrary` impl (proptest).

plafer reviewed (commented):
Sorry for the late review! Left mostly nits and questions

bobbinth commented on assembly/src/ast/procedure/name.rs:
```diff
@@ -328,3 +333,47 @@ impl FromStr for ProcedureName {
+impl Deserializable for ProcedureName {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let str: String = source.read()?;
+        let proc_name =
+            ProcedureName::new_unchecked(Ident::new_unchecked(Span::unknown(Arc::from(str))));
```
This would allow deserialization of arbitrary strings as procedure names, right? (e.g., strings with control or unprintable characters etc.) Would this not be a problem?

Should we enforce at least some kind of validation (e.g., maybe against the alphabet used on line 371 below)?

greenhat replied:
Good point! Since it's an unchecked user input, I suppose there is a chance to use it for some sort of attack. I cannot come up with any, but I'd feel safer if we have at least some validation here. This should be a different (more relaxed) validation from the one we use when parsing the MASM source code. @bitwalker What do you think?

greenhat replied:
I relaxed the validation rules in the `Ident` (used in `LibraryPath`) and `ProcedureName` in `new` constructors and used them in the deserialization instead of the `new_unchecked` ones.
The validation on MASM source code parsing is implemented in the lexer and does not use the checked constructors (it calls `new_unchecked`). So the existing checked constructors __are__ the different (from lexer's) kind of validation to be used everywhere else.

bobbinth commented on assembly/src/library/path.rs:
```diff
@@ -505,7 +506,26 @@ impl Deserializable for LibraryPath {
+        let path = LibraryPath::new(path).unwrap_or_else(|_| {
+            // Try to parse at least the namespace
+            match LibraryNamespace::strip_path_prefix(path) {
+                Ok((ns, rest)) => {
+                    let module_id = Ident::new_unchecked(Span::new(
+                        SourceSpan::default(),
+                        Arc::from(rest.to_string()),
+                    ));
+                    LibraryPath::new_from_components(ns, [module_id])
+                },
+                Err(_) => {
+                    let module_id = Ident::new_unchecked(Span::new(
+                        SourceSpan::default(),
+                        Arc::from(path.to_string()),
+                    ));
+                    LibraryPath::new_from_components(LibraryNamespace::Anon, [module_id])
+                },
```
Similar comment as above. I think we discussed this previously, but I don't remember why we are skipping validation rather than updating validation rules to support the newly required formats.

greenhat replied:
AFAIR, the reason was that validation that we have now is for the MASM source code parsing, and technically, we can allow any characters in the identifiers in the compiled code. @bitwalker correct me if I'm wrong. 
But from the "validate any user input" point of view, it makes sense to create and use another, more relaxed type of validation here and in `ProcedureName` (see my comment above).

greenhat replied:
Done. See my comment above - https://github.com/0xPolygonMiden/miden-vm/pull/1544#discussion_r1924733415

greenhat commented on package/src/package.rs:
```diff
@@ -0,0 +1,305 @@
+    pub digest: Digest,
+    // Signature will be added in the future when the type signatures are available in the Assembly
+    // #[serde(default)]
+    // pub signature: Option<Signature>,
```
Thanks! Done in https://github.com/0xPolygonMiden/miden-vm/pull/1544/commits/7e37745e88dba8c048a2e3447282a81ba5a53691

greenhat commented on package/src/package.rs:
```diff
@@ -0,0 +1,305 @@
+            .field("name", &format_args!("{}", self.name))
+            .field("digest", &format_args!("{}", DisplayHex::new(&self.digest.as_bytes())))
+            // .field("signature", &self.signature)
+            .finish()
```
Neat! Thanks! Done in https://github.com/0xPolygonMiden/miden-vm/pull/1544/commits/7e37745e88dba8c048a2e3447282a81ba5a53691

greenhat commented on package/README.md:
```diff
@@ -0,0 +1,12 @@
+- "MASP" magic bytes (4 bytes);
+- Version of the package binary format (3 bytes, e.g. "1.0");
+
+The header is followed by the `Package` serialized with `serde` using the [`bitcode`](https://docs.rs/bitcode/latest/bitcode/#structs) encoder to minimize the size.
```
I believe `bitcode` was chosen earlier in the compiler repo `Package` implementation for producing the smallest serialized size (see [benchmark](https://github.com/djkoloski/rust_serialization_benchmark)) and compatibility with `serde`. Since the largest part of the `Package` (`MastForest`) is not serialized with `bitcode` the serialized size reason probably does not work that much.
I admit, the downsides of `bitcode` being Rust-only and not a stable format are pretty significant. Not to mention the inconsistency with the established serialization approach in the codebase.
I'm ready to rework this PR and get rid of `bitcode` in favor of manual `Serializable/Deserializable` implementation.

@bitwalker What's your opinion?

bobbinth replied:
> I believe `bitcode` was chosen earlier in the compiler repo `Package` implementation for producing the smallest serialized size (see [benchmark](https://github.com/djkoloski/rust_serialization_benchmark)) and compatibility with `serde`.

I would be surprised it it is more compact than manual implementation of `Serializable`/`Deserializable` traits - but also, I'm not really user if for packages, the size is of primary importance as we won't put packages on chain. Only `MastForest`s will go onchain and we still need to implement a more compact way of serializing `MastForest`s (the current implementation is optimized for deserialization speed rather than file size size).

We haven't really tried to provide compatibility with `serde` for most things (main exceptions that come to mind are config files for which we usually use TOML) - so, I'm thinking we could probably drop this requirement from packages too.

greenhat replied:
I agree and will switch to `Serializable/Deserializable`.
 
On the compact serialized size note.`bitcode` uses integer packing to avoid serializing "empty" parts. We can use some [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) encoding in our `ByteWriter/ByteReader` as well (e.g. store `1u32` in one byte). I had a positive experience with Varint (+zigzag for signed values) encoding in the past.

bobbinth replied:
`ByteWriter/ByteReader` use variable length encoding ([vint64](https://docs.rs/vint64/latest/vint64/)) for `usize` values (see [here](https://github.com/facebook/winterfell/blob/main/utils/core/src/serde/byte_writer.rs#L77)). But `u32`, `u64` etc. are serialized as their respective number of bytes.

greenhat replied:
Oh, I missed the `usize` encoding. Thanks! Why not use variable length encoding for `u32` and `u64`?

bobbinth replied:
Fixed-length encoding was left for these to make de-serialization as simple (and as fast) as possible. Though, I don't think we've measured the real-world impact of the difference - so, may be an over-optimization.

greenhat replied:
Got it. My motto for any on-chain data is "every bit counts". Optimizing for speed later is easier than changing the binary format.

bobbinth replied:
Agreed - though the current format would not go on chain (i.e,. we probably won't put packages on chain in their current form).

bobbinth commented on package/src/dep/mod.rs:
```diff
@@ -0,0 +1,71 @@
+/// The name of a dependency
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub enum DependencyName {
+    /// The dependency is a system library
+    System(SystemLibraryId),
+    /// The dependency is a user library with the given name
+    User(String),
+}
```
Related to the previous comment, maybe this could be just `DependencyName(String)`?

greenhat replied:
Done.

bobbinth reviewed (commented):
Looks good! Thank you! I left some comments inline. The main ones are about:

- Potentially adding some validation to library path deserialization.
- Using `Serializable`/`Deserializable` instead of `serde`.
- Not including functionality that assumes downstream crates (i.e., `miden-base`).

greenhat commented on package/src/package.rs:
```diff
@@ -0,0 +1,305 @@
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct PackageExport {
+    /// The fully-qualified name of the procedure exported by this package
+    pub name: String,
```
Good question! AFAIR, the reason was probably that Wasm CM naming does not fit in `QualifiedProcedureName` well (e.g. `namespace:package/interface@version#function`). But we're squeezing Wasm CM names in `QualifiedProcedureName` in `Library::exports` anyway (under `anon` namespace). So I see no reason for `String` here. Will switch to `QualifiedProcedureName`.

greenhat replied:
Done.

greenhat commented on package/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "miden-package"
```
I got so much used to `miden-package` so I'm biased here. :) In general, if we plan to have another "packages" (account?) giving it a more specific name makes sense.

bobbinth replied:
What do you think about `miden-mast-package` here? I would also maybe expand the description a little bit to make it more specific.

greenhat replied:
Works for me! With MAST being the core part of the package, it makes sense to put `mast` in the name. I'll rename it and expand the description.

greenhat replied:
Done.

greenhat commented on package/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+name = "miden-package"
+version = "0.11.0"
+description = "Miden VM package"
+documentation = "https://docs.rs/miden-package/0.10.5"
```
Thanks. Done.

greenhat commented on package/README.md:
Thanks. Done.

greenhat commented:
@bobbinth @plafer Thank you for the review! I've addressed all notes, implemented easy ones, and I am converting this PR to draft to work on:
- Switch `PackageExport::name` to have `QualifiedProcedureName` type and propagate `Arbitrary` implementation for `proptest`;
- Remove `bitcode` and `serde` based serialization in favor of `Serializable/Deserializable`;
- New validation for ids (`LibraryPath`, `ProcedureName`) on deserialization;

bobbinth commented on package/src/lib.rs:
```diff
@@ -0,0 +1,27 @@
+#[cfg(test)]
+mod tests;
+
+pub use self::{
+    dep::{
+        resolver::{
+            DependencyResolver, LocalResolvedDependency, MemDependencyResolverByDigest,
+            ResolvedDependency,
+        },
+        Dependency, DependencyName,
+    },
+    package::{MastArtifact, Package, PackageExport, PackageManifest},
+};
```
Question: should we re-export such things as `Digest`, `Program`, `MastForest` etc. from here is that this crate is more "self-contained"?

greenhat replied:
I'd say "yes". I can imagine `miden-package` crate being used for some sort of package analysis without relaying on the rest of VM crates.

greenhat replied:
Done.

bobbinth commented on package/src/package.rs:
```diff
@@ -0,0 +1,296 @@
+    // const MAGIC: &'static [u8] = b"MASP\0";
+    // const FORMAT_VERSION: &'static [u8] = b"1.0\0";
+
+    ///// Parses a package from the provided bytes
+    //pub fn read_from_bytes<B>(bytes: B) -> Result<Self, Report>
+    //where
+    //    B: AsRef<[u8]>,
+    //{
+    //    use alloc::borrow::Cow;
+    //
+    //    let bytes = bytes.as_ref();
+    //
+    //    let bytes = bytes
+    //        .strip_prefix(Self::MAGIC)
+    //        .ok_or_else(|| Report::msg("invalid package: missing header"))?;
+    //    let bytes = bytes.strip_prefix(Self::FORMAT_VERSION).ok_or_else(|| {
+    //        Report::msg(format!(
+    //            "invalid package: incorrect version, expected '1.0', got '{}'",
+    //            bytes.get(0..4).map(String::from_utf8_lossy).unwrap_or(Cow::Borrowed("")),
+    //        ))
+    //    })?;
+    //
+    //    bitcode::deserialize(bytes).map_err(Report::msg)
+    //}
+    //
+    ///// Serializes the package into a byte array
+    //pub fn write_to_bytes(&self) -> Result<Vec<u8>, Report> {
+    //    let mut bytes = Vec::new();
+    //    bytes.extend_from_slice(Self::MAGIC);
+    //    bytes.extend_from_slice(Self::FORMAT_VERSION);
+    //    let mut data = bitcode::serialize(self).map_err(Report::msg)?;
+    //    bytes.append(&mut data);
+    //    Ok(bytes)
+    //}
```
Is this meant to be commented-out?

greenhat replied:
Yes, I missed it. Thanks for spotting this! 
In my defense, I planned to give this PR a "polish" before marking it ready and asking for another round of reviews. 🙂

greenhat replied:
Done.

bobbinth commented on package/src/package.rs:
```diff
@@ -0,0 +1,296 @@
+    //    Ok(bytes)
+    //}
+
+    pub fn digest(&self) -> Digest {
+        self.mast.digest()
+    }
```
nit: let's add a doc comment here.

greenhat replied:
Done.

bobbinth commented on package/src/tests.rs:
Should this test be moved under `src/package/serialization`?

greenhat replied:
Done.

bobbinth commented on package/src/package.rs:
Should this file be moved to `src/package/mod.rs`?

greenhat replied:
Done.

bobbinth commented on package/src/package.rs:
```diff
@@ -0,0 +1,296 @@
+/// A package containing a [Program]/[Library], and a manifest (exports and dependencies).
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct Package {
+    /// Name of the package
+    pub name: String,
+    /// The MAST artifact ([Program] or [Library]) of the package
+    pub mast: MastArtifact,
+    /// The package manifest, containing the set of exported procedures and their signatures,
+    /// if known.
+    pub manifest: PackageManifest,
+}
```
Probably not for this PR, but where would we put custom metadata? Here, I'm thinking about the storage layout metadata we'd need to instantiate account components.

greenhat replied:
I'm thinking, either directly in `Package` or in `PackageManifest`.

bobbinth replied:
Let's create an issue for this.

greenhat replied:
Done - https://github.com/0xPolygonMiden/miden-vm/issues/1662

bobbinth reviewed (commented):
Looks good! Thank you! I left some more comments/questions inline.

Besides these, I think the main thing remaining here is potentially adding some validation to library path deserialization - right?

greenhat commented on package/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+serde_bytes = "0.11.15"
+serde_repr = "0.1.19"
+bitcode = { version = "0.6.3", default-features = false, features = ["serde"] }
+derive_more = "0.99"
```
Done.

greenhat commented:
@bobbinth @plafer I implemented everything and addressed all your notes. Please do another round.

plafer reviewed (approved):
LGTM!

bobbinth commented on assembly/src/ast/procedure/name.rs:
```diff
@@ -311,10 +333,12 @@ impl FromStr for ProcedureName {
                 }
             },
-            Some((_, c)) if c.is_ascii_lowercase() || c == '_' || c == '$' => {
+            Some((_, c))
+                if c.is_ascii_lowercase() || c == '_' || c == '-' || c == '$' || c == '.' =>
+            {
```
Not for this PR, but the code in this function is a bit difficult to follow. Let's create an issue (could be a "good first issue") to make the code more readable.

greenhat replied:
I agree. I made https://github.com/0xPolygonMiden/miden-vm/issues/1663 for it.

bobbinth commented on assembly/src/library/path.rs:
```diff
@@ -505,7 +505,23 @@ impl Deserializable for LibraryPath {
             str::from_utf8(path).map_err(|e| DeserializationError::InvalidValue(e.to_string()))?;
-        Self::new(path).map_err(|e| DeserializationError::InvalidValue(e.to_string()))
+        LibraryPath::new(path).or_else(|_| {
+            // Try to parse at least the namespace
```
Could we expand the comment a bit here to explain what the code below is intending to cover? My understanding is:

If `path` is not a valid `LibraryPath`, we try to take the first "leg" of the path and use it as a namespace and the rest of the string as the module name. And if there is no first "leg", then we use `Anon` namespace and try to use the whole `path` as the module name.

If the above is correct, it is not clear to me what exactly we are trying to enable here. It seems like maybe we want to make sure that a path with no namespaces get `Anon` namespace? If so, could we simplify the code as:
```rust
LibraryPath::new(path).or_else(|_| {
    let module_id = Ident::new(path).map_err(|e| {
        DeserializationError::InvalidValue(format!("Invalid module id: {}", e))
    })?;
    Ok(LibraryPath::new_from_components(LibraryNamespace::Anon, [module_id]))
}
```

greenhat replied:
Good point! This fallback code was written before I relaxed the `Ident` validation (module names). Now, we parse Wasm CM names fine, and we don't need any fallback, so I removed it in https://github.com/0xPolygonMiden/miden-vm/pull/1544/commits/a0200635c32c0fc90f2786f6fef0fdd200a5aec3

bobbinth commented on package/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "miden-package"
+version = "0.11.0"
+description = "Miden VM package"
+documentation = "https://docs.rs/miden-package/0.11.0"
```
Let's increment the versions to v0.12.0.

greenhat replied:
Thanks! Done.

bobbinth reviewed (commented):
Thank you! Looks good! I left some more comments inline - most are pretty minor.

bobbinth commented:
Let's also resolve the merge conflicts.

greenhat commented on package/src/package/serialization/mod.rs:
```diff
@@ -0,0 +1,207 @@
+            Library::read_from(source).map(Arc::new).map(MastArtifact::Library)
+        } else {
+            Err(DeserializationError::InvalidValue(format!(
+                "Invalid MAST artifact tag: {:?}",
```
Thanks! Done.

bobbinth replied:
It still seems to be capitalized in the code.

greenhat commented on package/src/package/mod.rs:
```diff
@@ -0,0 +1,262 @@
+    pub dependencies: Vec<Dependency>,
+}
+
+/// A procedure exported by a package, along with its digest and
+/// signature(will be added after MASM type attributes are implemented).
```
Done.

greenhat commented on package/src/lib.rs:
```diff
@@ -0,0 +1,25 @@
+#[cfg(test)]
+extern crate std;
+
+pub use vm_core::{chiplets::hasher::Digest, mast::MastForest, Program};
```
Thank you! I missed them. Done.

greenhat commented:
@bobbinth Thank you for the review! Thank you for your patience. I addressed all the notes, renamed the crate and rebased the branch. Please do another round.

bobbinth reviewed (approved):
Looks good! Thank you! I left of very minor nit inline and also the change from https://github.com/0xPolygonMiden/miden-vm/pull/1544#discussion_r1929614558 somehow got reverted.

Once the above is done, should we merge it into `next`?

greenhat commented on package/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+serde_bytes = "0.11.15"
+serde_repr = "0.1.19"
+bitcode = { version = "0.6.3", default-features = false, features = ["serde"] }
+derive_more = "0.99"
```
Thank you! Done. 
I probably botched this part during the rebase (we don't need serde anymore) so I cleaned it up.

greenhat commented:
> Looks good! Thank you! I left of very minor nit inline and also the change from [#1544 (comment)](https://github.com/0xPolygonMiden/miden-vm/pull/1544#discussion_r1929614558) somehow got reverted.

Thanks! Fixed.

> Once the above is done, should we merge it into `next`?

Yes, we can merge it into `next`. I branched out https://github.com/0xPolygonMiden/miden-vm/tree/greenhat/miden-package-pre-eam-vm-for-compiler from this branch before the rebase to use the pre-element-addressable-memory version of this branch in the compiler until we migrate our memory model.

bobbinth reviewed (approved):
All looks good! Thank you!
